cajita/src/Cajita_SparseIndexSpace.hpp

```diff
@@ -0,0 +1,499 @@
+        return bit_length( N >> 1 ) + static_cast<Integer>( 1 );
+    else
+        return 0;
+}
+
+template <typename Integer>
+constexpr Integer bit_count( Integer N ) noexcept
+{
+    return bit_length( N - 1 );
+}
+
+template <typename Integer>
+constexpr Integer binary_reverse( Integer data,
+                                  char loc = sizeof( Integer ) * 8 - 1 )
+{
+    if ( data == 0 )
+        return 0;
+    return ( ( data & 1 ) << loc ) | binary_reverse( data >> 1, loc - 1 );
+}
+
+template <typename Integer>
+constexpr unsigned count_leading_zeros( Integer data )
+{
+    unsigned res{0};
+    data = binary_reverse( data );
+    if ( data == 0 )
+        return sizeof( Integer ) * 8;
+    while ( ( data & 1 ) == 0 )
+        res++, data >>= 1;
+    return res;
+}
+
+constexpr int bit_pack( const uint64_t mask, const uint64_t data )
+{
+    uint64_t slresult = 0;
+    uint64_t &ulresult{slresult};
+    uint64_t uldata = data;
+    int count = 0;
+    ulresult = 0;
+
+    uint64_t rmask = binary_reverse( mask );
+    unsigned char lz{0};
+
+    while ( rmask )
+    {
+        lz = count_leading_zeros( rmask );
+        uldata >>= lz;
+        ulresult <<= 1;
+        count++;
+        ulresult |= ( uldata & 1 );
+        uldata >>= 1;
+        rmask <<= lz + 1;
+    }
+    ulresult <<= 64 - count; // 64 bits (maybe not use a constant 64 ...?)
+    ulresult = binary_reverse( ulresult );
+    return (int)slresult;
+}
+
+constexpr uint64_t bit_spread( const uint64_t mask, const int data )
+{
+    uint64_t rmask = binary_reverse( mask );
+    int dat = data;
+    uint64_t result = 0;
+    unsigned char lz{0};
+    while ( rmask )
+    {
+        lz = count_leading_zeros( rmask ) + 1;
+        result = result << lz | ( dat & 1 );
+        dat >>= 1, rmask <<= lz;
+    }
+    result = binary_reverse( result ) >> count_leading_zeros( mask );
+    return result;
+}
+//---------------------------------------------------------------------------//
+
+// Tile Hash Key (hash) <=> TileID
+template <typename Key, typename HashType>
+struct TileID2HashKey;
+
+template <typename Key, typename HashType>
+struct HashKey2TileID;
+
+// functions to compute hash key
+// can be rewriten in a recursive way
+template <typename Key>
+struct TileID2HashKey<Key, HashTypes::Naive>
+{
+    TileID2HashKey( std::array<int, 3> &&tnum )
+        : _tilenum( tnum )
+    {
+    }
+    TileID2HashKey( int i, int j, int k )
+        : _tilenum( i, j, k )
+    {
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    constexpr auto operator()( std::array<int, 3> tileid ) -> Key
+    {
+        return tileid[0] * _tilenum[1] * _tilenum[2] + tileid[1] * _tilenum[2] +
+               tileid[2];
+    }
+
+    KOKKOS_FORCEINLINE_FUNCTION
+    constexpr auto operator()( int tileid_x, int tileid_y, int tileid_z ) -> Key
+    {
+        return tileid_x * _tilenum[1] * _tilenum[2] + tileid_y * _tilenum[2] +
+               tileid_z;
+    }
+
+  private:
+    std::array<int, 3> _tilenum;
```

Comment:
store this as either 3 separate integers or in a `Kokkos::Array` so we can trivially copy to and from the device